opengovsg/FormSG
# feat: add otp verification to payment contact field

## Problem

Email fields for payment currently do not verify emails before accepting them to be tagged to payments. This may be an issue when we need to reconcile payments and there is no identifier for us to match payments to users. Verified emails provide us a way to do so.

Closes #5964

## Solution

Modify existing email verification mechanisms in form fields to use for payments.

**Breaking Changes** 



**Features**:

- OTP verification on payment contact email field! 
- ... and tests! 

## Before & After Screenshots

**BEFORE**:

(taken from https://github.com/opengovsg/FormSG/pull/5917, notice the lack of a "Verify" button beside the email field)

https://user-images.githubusercontent.com/37061143/229749350-82839a58-d98a-4792-9950-c8c18617b39e.mp4

**AFTER**:


https://user-images.githubusercontent.com/37061143/229748863-397bfce7-f5a7-4826-8fd1-5e24341a3b40.mov

## Tests


To test the expected behaviour on payment enabled form after feature has been implemented (on a payment form with no other verifiable fields):
1. - [ ] If payment contact field is not filled in, user should not be able to proceed with payment.
2. When “Verify” in any verifiable email field component in the form is pressed,
  - [ ] In the DB, a transaction should be created in the `verifications` table. 
    - [ ] If payment is enabled on the form, this transaction should include the `paymentField` key with the default values of `IVerificationFieldSchema` if the verifiable email field to be verified is not the payment one. 
    - [ ] If it is the payment verifiable email field, the value stored in `paymentField` key should be `IVerificationFieldSchema` but with the values populated according to the otp hash calculated.
  - [ ] On the frontend, the verification modal should open up for the user to key in the OTP.
  - [ ] An email with the OTP should be sent to the entered email.
3. Press "Resend OTP" on the payment contact email field once possible (usually after 60s).
  - [ ] OTP-prefix should be updated accordingly.
  - [ ] A new transaction should be added to the `verifications` table in the DB.
  - [ ] An email with the new OTP should be sent to the entered email.
4. Enter a wrong OTP in the payment contact field and click "Submit".
  - [ ] The verification box should let the user know that the OTP is wrong.
5. Enter the appropriate OTP in the payment contact field and click "Submit".
  - [ ] The verification box should collapse and show that the email is verified as per normal verifiable email fields.
6. - [ ] Proceed to pay in the payment element. If the app is given the live stripe credentials and the payment is successful, a stripe payment receipt should be sent to the entered email.

- [ ] Repeat tests 1-6 for a payment form with other verifiable fields.
- [ ] Repeat tests 2-6 for verifiable fields in a storage non-payment form.
- [ ] Repeat tests 2-6 for verifiable fields in an email form.
- [ ] Repeat steps 2-6 on a storage form with both payment and 2 verifiable fields in the form (1 email + 1 mobile)
